### PR TITLE
GH-96163: speed up `random.randrange`

### PR DESCRIPTION
**Pull request by matthiasgoergens**
This change speeds up the most common code path through
`_randbelow_with_getrandbits`.

That improves the execution speed of functions like `random.randrange`.

See https://github.com/python/cpython/issues/96163 for benchmarks and a
discussion.

* Issue: gh-96163

### CODE REVIEW

**Comment by matthiasgoergens** [comment]
@rhettinger I see that you have been automatically assigned as a reviewer.  Please don't feel the need to hurry with your review and have a look whenever it's convenient for you.

The speed improvement itself is rather small in absolute terms.  See the benchmarks in the linked issue.  (Btw, I was very surprised to learn how much variance there is in running the same code multiple times and measuring timing!)

I thought this might be worthwhile regardless, because you impressed on my how important fast execution is for these functions.

I presented two code alternatives in the linked issue.  This is the one that improves speed a bit more, but the code is also slightly more complicated.  I'm more than happy to go with the ever so slightly slower but cleaner one, too.

**Comment by rhettinger** [comment]
-1 To my eyes, these revisions look gross and will hinder future maintenance. I strongly prefer current code which is much cleaner and expresses what it is doing in a straightforward manner.

File: Lib/random.py [hunk 1]
```diff
@@ -300,19 +308,21 @@ def randrange(self, start, stop=None, step=_ONE):
...
         width = istop - istart
         istep = _index(step)
         # Fast path.
         if istep == 1:
-            if width > 0:
+            try:
                 return istart + self._randbelow(width)
-            raise ValueError(f"empty range in randrange({start}, {stop})")
+            except ValueError as e:
+              raise ValueError(f"empty range in randrange({start}, {stop})") from e
```
Comment (Fidget-Spinner) [inline Lib/random.py]: Just so you're aware: In 3.11, try-except paths are near zero cost for the happy path (no exception raised), but they are **slower** in the case of an exception being raised. So this `ValueError` shouldn't be a common thing!

**Comment by matthiasgoergens** [comment]
> -1 To my eyes, these revisions look gross and will hinder future maintenance. I strongly prefer current code which is much cleaner and expresses what it is doing in a straightforward manner.
> 
> Also, the interpreter implementation is undergoing rapid change of late. We don't want to chase local minimums or tune to a single build.

Ok, that's all fair. In this case, please have a look at https://github.com/matthiasgoergens/cpython/pull/8/commits/a0e22e4c30876e4482e5adf496335d6785425c15 which simplifies and speeds up (just slightly less so than the current state of this PR.)

File: Lib/random.py (hunk 1 above)
Comment (matthiasgoergens) [reply Lib/random.py]: Yes, I Benchmarked everything against 3.11 as downloaded from the Python website for Mac, and against a local optimised build of 'main'. I was surprised how cheap try-except blocks are when they don't catch anything.

**Comment by serhiy-storchaka** [comment]
My first reaction was strictly negative, but reading the issue I see that your initial proposition was more moderate. The current code includes a micro-optimization, which was added because method resolution in Python was relatively slow, in comparison with saving and loading a local variable. But method call was severely optimized several versions ago. This micro-optimization has now much lesser benefit if any. It is not first time when we get rid of old micro-optimizations for sake of clearer code, even at the cost small performance loss. And it seems that in this case we will get a performance gain.

@matthiasgoergens, could you please test your initial proposition with older Python versions: 3.10, 3.6 (or earlier) and something old, close to 2.4 (when this micro-optimization was added).

I would write the code without using the `:=` operator:
```python
        while True:
            r = self.getrandbits(n.bit_length())
            if r < n:
                return r
```
But it is a matter of style, I don't expect any difference in performance. Just for the case, could you also compare the two variants on 3.11?

In any case, the decision is Raymond's.

As for the monstrous patch, I see reasons of every change, they make sense to me, I like them in separation, but I think that we should stop here. This takes us too far, all together makes the code unreadable, and the benefit is less and less, it's not worth it.

**Comment by matthiasgoergens** [comment]
> @matthiasgoergens, could you please test your initial proposition with older Python versions: 3.10, 3.6 (the method call optimizations were added in 3.7 and later) and something old, close to 2.4 (when this micro-optimization was added).

Yes, I'll do those tomorrow.

I changed the PR to reflect the simpler suggestion.  I did a force-posh, because this simpler suggestion doesn't build on the more complicated one at all, so should be reviewed from scratch anyway.

> I would write the code without using the `:=` operator:
> 
> ```python
>         while True:
>             r = self.getrandbits(n.bit_length())
>             if r < n:
>                 return r
> ```
> 
> But it is a matter of style, I don't expect any difference in performance. Just for the case, could you also compare the two variants on 3.11?

I compared both version on 3.11, here's the graph:

serhiy

Interestingly enough, the Walrus seems to be ever so slightly faster.  (But I am a bit tired, so I'll double check tomorrow.)

Using `dis.dis` on both versions shows that the Walrus versions seems to have almost twice as much Bytecode as your suggestion.

The Walrus version disassembled:

```
>>> dis.dis(random2.Random()._randbelow)
238           0 RESUME                   0

241           2 LOAD_FAST                0 (self)
              4 LOAD_METHOD              0 (getrandbits)
             26 LOAD_FAST                1 (n)
             28 LOAD_METHOD              1 (bit_length)
             50 PRECALL                  0
             54 CALL                     0
             64 PRECALL                  1
             68 CALL                     1
             78 COPY                     1
             80 STORE_FAST               2 (r)
             82 LOAD_FAST                1 (n)
             84 COMPARE_OP               5 (>=)
             90 POP_JUMP_FORWARD_IF_FALSE    46 (to 184)

242     >>   92 NOP

241          94 LOAD_FAST                0 (self)
             96 LOAD_METHOD              0 (getrandbits)
            118 LOAD_FAST                1 (n)
            120 LOAD_METHOD              1 (bit_length)
            142 PRECALL                  0
            146 CALL                     0
            156 PRECALL                  1
            160 CALL                     1
            170 COPY                     1
            172 STORE_FAST               2 (r)
            174 LOAD_FAST                1 (n)
            176 COMPARE_OP               5 (>=)
            182 POP_JUMP_BACKWARD_IF_TRUE    46 (to 92)

243     >>  184 LOAD_FAST                2 (r)
            186 RETURN_VALUE
```

Your non-Walrus style version disassembled:

```
>>> dis.dis(random3.Random()._randbelow)
238           0 RESUME                   0

241           2 NOP

242     >>    4 LOAD_FAST                0 (self)
              6 LOAD_METHOD              0 (getrandbits)
             28 LOAD_FAST                1 (n)
             30 LOAD_METHOD              1 (bit_length)
             52 PRECALL                  0
             56 CALL                     0
             66 PRECALL                  1
             70 CALL                     1
             80 STORE_FAST               2 (r)

243          82 LOAD_FAST                2 (r)
             84 LOAD_FAST                1 (n)
             86 COMPARE_OP               0 (<)
             92 POP_JUMP_FORWARD_IF_FALSE     2 (to 98)

244          94 LOAD_FAST                2 (r)
             96 RETURN_VALUE

241     >>   98 JUMP_BACKWARD           48 (to 4)
```

If I read the disassembly right, it looks like the Walrus avoids one `LOAD_FAST` of `r`?

Style-wise, I don't really have a strong opinion either way.  I just picked the Walrus, because it looks almost like the textbook examples suggested in https://peps.python.org/pep-0572/

> As for the monstrous patch, I see reasons of every change, they make sense to me, I like them in separation, but I think that we should stop here. This takes us too far, all together makes the code unreadable, and the benefit is less and less, it's not worth it.

Agreed.

Yes, the original proposal was rather modest.

I blame Tim Peters encouragement for the monster, when he said, "However, I think this needs more fiddling." ;-)

**Comment by matthiasgoergens** [comment]
@serhiy-storchaka If you like https://github.com/matthiasgoergens/cpython/pull/10/commits/f8f9f4a6a5e69482fbf70a6e7c9f128ae7ee8dab we can remove other obsolete micro-optimizations, too.

**Comment by serhiy-storchaka** [comment]
> @serhiy-storchaka If you like [matthiasgoergens@f8f9f4a](https://github.com/matthiasgoergens/cpython/commit/f8f9f4a6a5e69482fbf70a6e7c9f128ae7ee8dab) we can remove other obsolete micro-optimizations, too.

It is not microoptimization, it is a matter of style. If you have no evidences that these changes make the code faster, I do not see reasons of changing it.

**Comment by serhiy-storchaka** [comment]
> I compared both version on 3.11, here's the graph:

It is surprising that the difference between walrus and non-walrus variants is such large and depends on `n`. It is suspicious that the walrus variant is always beats the baseline on that graph, while on previous graphs it was not so for `n` close to the power of two. Are you sure that it was the only difference?

While we are here, please try also to compare with the code that does not inline `k`, but calculates `n.bit_length()` only once outside of the loop. The effect of caching `getrandbits` depends on the cost of the method call, so the effect of this optimization may be changed with time. But it is not so for caching `n.bit_length()`.

**Comment by matthiasgoergens** [comment]
I ran some more benchmarks. But not enough time to finish the write-up for them today.

The discrepancy you noted between the last graph and the previous graphs is actually a difference between 'main' and 3.11.

In 3.11 the simple proposal is indeed always better than what's in the standard library. That's what I happened to use for the last graph.

The previous graphs used main, and there the standard library can hold its own in some special cases against the proposal.

However both are outperformed by the variation of the proposal that uses 'k' to store the bit length.

**Comment by matthiasgoergens** [comment]
I finally have my desktop computer back.  (It's motherboard had died.)  The benchmarks looks a lot less noisy on the desktop (running Archlinux) than on the laptop running (MacOS).

> It is suspicious that the walrus variant is always beats the baseline on that graph, while on previous graphs it was not so for n close to the power of two. Are you sure that it was the only difference?

Well spotted.  I was suspicious about that one too.  The big difference was that the previous graphs showed the benchmarks from 'main', and for that one I had randomly picked 3.11.  I hadn't expected such a difference.

In any case, here are new benchmarks run on my desktop.

## Python 3.6

randbelow_36

In Python 3.6 the stdlib doesn't have the 'static' switch between the different option.  You can see that this is slow.

Overwriting `_randbelow` with something that just always does what the current standard library does, worked best.  Confirming that the micro-optimization we find in the current code was actually an optimization at some point.

```python
class Random_vanilla(random.Random):
    def _randbelow(self, n):
        "Return a random int in the range [0,n).  Defined for n > 0."

        getrandbits = self.getrandbits
        k = n.bit_length()
        r = getrandbits(k)  # 0 <= r < 2**k
        while r >= n:
            r = getrandbits(k)
        return r
```

## Python 3.11

I used Archlinux's aura to install the latest release candidate of `3.11.0rc1`.

randbelow_311-a

This one is a bit messy, because I tried to cram everything in.  What you can hopefully see is that I can confirm that in 3.11 the initial clean proposal beats the status quo even just above powers of 2.

Tied for best performers we have the two variants that materialize ` k = n.bit_length()`:

```python
class Random_bit_length(random.Random):
    def _randbelow(self, n):
        "Return a random int in the range [0,n).  Defined for n > 0."

        k = n.bit_length()
        # 0 <= r < 2**k
        while (r := self.getrandbits(k)) >= n:
            pass
        return r

class Random_k(random.Random):
    def _randbelow(self, n):
        k = n.bit_length()
        while True:
            r = self.getrandbits(k)
            if r < n:
                return r
```

Using the Walrus or not doesn't make a difference in performance in 3.11.

## Python 3.12

I compiled current 'main' with optimizations.

randbelow_312-b

Compared to 3.11 the code that's in the standard library improved quite a bit.  I suspect something changed that made making a local variable cheaper?

The pattern of relative strengths between the approaches is the same as the graphs I showed you at first.  Just with much less noise in the data, because my desktop PC runnig Linux is a better machine for benchmarking than my Laptop was.

The clear performance winner is this variant of the proposal you suggested:

```python
class Random_bit_length(random.Random):
    def _randbelow(self, n):
        "Return a random int in the range [0,n).  Defined for n > 0."

        k = n.bit_length()
        # 0 <= r < 2**k
        while (r := self.getrandbits(k)) >= n:
            pass
        return r
```

I wonder what changed in CPython, because now using the Walrus operator beats out the alternative loop version you provided; but they were tied in 3.11.

**Comment by matthiasgoergens** [comment]
@serhiy-storchaka About the other simplification proposal:

Sorry, I mixed in a few changes without explanation.

First: I removed all local variables of the form `random = self.random`.

Second: many variables were imported with an underscore in front, just so that the code can create local variables without the underscore:

```python
-from bisect import bisect as _bisect
+from bisect import bisect
```

```python
@@ -415,30 +412,28 @@ def sample(self, population, k, *, counts=None):
             if total <= 0:
                 raise ValueError('Total of counts must be greater than zero')
             selections = self.sample(range(total), k=k)
-            bisect = _bisect
             return [population[bisect(cum_counts, s)] for s in selections]

```

If you think that's just a style question, an alternative would be to leave the import as is, but change the code like this:
```python
@@ -415,30 +412,28 @@ def sample(self, population, k, *, counts=None):
             if total <= 0:
                 raise ValueError('Total of counts must be greater than zero')
             selections = self.sample(range(total), k=k)
-            bisect = _bisect
-            return [population[bisect(cum_counts, s)] for s in selections]
+            return [population[_bisect(cum_counts, s)] for s in selections]
```

(I suspect the underscore imports were made to clearly differentiate between local variables and imports, because that used to be an important performance consideration.)

Obviously, performance doesn't care whether variable names have an underscore or not.  But not making local variables is simpler, and now also faster.

**Comment by serhiy-storchaka** [comment]
Great work! Thank you, @matthiasgoergens.

I think that we can now only consider variants with `k = n.bit_length()`.

As for difference between walrus and non-walrus variants, the only meaningful difference is the NOP operation. I think it is a compiler regression. In older versions NOPs were only added by the peephole optimizer after unconditional jumps, so they never were executed. But now they are added at the execution path.

Other proposals should be discussed separately. I am not sure that that caching a global as a local in `sample()` makes any significant difference, because the cost of calling function is usually much larger than the cost of reading a global, but it's all on Raymond.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

@rhettinger, your turn.

Seems, one of microoptimizations is no longer needed, and only adds an overhead. The code is clearer and even faster without it.

**Comment by rhettinger** [comment]
The walrus operator change shouldn't be done.  It reduces readability for no gain.  In 3.10, it replaces a STORE_FAST/LOAD_FAST pair with a DUP_TOP/STORE_FAST.  In 3.11, it is COPY/STORE_FAST.  For both versions, the walrus version is less readable and makes the generated word code slightly larger.

The other part of the edit was removing the local variable in favor a direct method lookup from *self*. This specific optimization was considered in another issue. It was tempting but was rejected. 

The gain was tied to recently added method caching. That gain was unstable and did not help on some builds. The caching works best in a timing loop but isn't always selected in real code (it depends on the call counts).  Also, the "win" was mainly confined to eliminating a single STORE_FAST, a cost which is only paid once (outside the loop).  After the first loop, there was no win and on some builds there was a loss.  My preference was keep the local variable optimization because it was stable across builds, it didn't chase the current state of the interpreter (which is still in flux).  Also, I chose to favor minimizing latency in the unhappy case (multiple calls to getrandbits()) rather than the average case (lucky first hit).  I considered non CPython implementations which all benefit from calling a boundmethod rather than rebinding on every call. Mostly, we write pure python code in a way that works well across builds and is resilient to upstream interpreter changes.   

One other note: the baseline timings are negatively affected by the increased cost for making a bound method.  Victor took out the caching for boundmethods which [increased their cost](https://docs.python.org/3.9/whatsnew/3.9.html#optimizations). That decision may yet be reversed because it slowed down code in cases like this.

Tim Peters gave us advice that applies here.  Don't chase the interpreter or a specific build into local minimums.  Think algorithmically and do things that make sense. He even favored accepting some changes that did less work algorithmically even though on a particular build they happened to make timings a little worse.  Applying his advice to the current code leads us to keep the [loop invariant code motion](https://en.wikipedia.org/wiki/Loop-invariant_code_motion) hoisting the method binding out of the loop rather than hoping the interpreter might do this for us.

Since this specific change was previously rejected, I'm going to mark this as closed. If you really feel strongly about relitigating this, go ahead and reopen this and we can spend more time debating on whether local variable optimizations should be kept because they are safe, stable, and conceptually the right thing to do, or whether we should save the cost of initial binding and store fast in hopes that callsite caching will always win across different call patterns, across builds, and across the ongoing updates to the interpreter (where presumably we would need to retime this will every release).

**Comment by Fidget-Spinner** [comment]
> I considered non CPython implementations which all benefit from calling a boundmethod rather than rebinding on every call.

Interestingly on CPython 3.11, that may not be true for the code here. The specialising interpreter caches unbound method calls such that they use zero lookups (no namespace lookup, no MRO hashtable lookup). The specialising interpreter does not cache bound method lookups.

That is to say
``a.b()``
May now be faster than
```
b = a.b
b()
```

On pypy I suspect the former might be faster too. However I understand Raymond's point on not chasing interpreter implementation detail.

**Comment by ericsnowcurrently** [comment]
Regardless of the outcome, @matthiasgoergens, thanks for taking the time to look into this!  There are many opportunities to improve performance that would benefit from the same attention to detail you showed here.

A key part of the issue is that the optimization is impacted by changes in the bytecode interpreter that are still in development and subject to change.  So subtle changes in syntax to take advantage of extra efficient branches in the interpreter may not pay off later as the interpreter changes.

**Pull request closed without merging**